spec/abci++/abci++_app_requirements_002_draft.md

```diff
@@ -177,3 +179,818 @@ Likewise, `ExtendVote` can also be non-deterministic:
+data. Likewise, the value of `Code` must be deterministic.
+Both the `Code` and `Data` are included in a structure that is hashed into the
+`LastResultsHash` of the block header in the next height (next block execution mode), or the
+header of the block to propose in the current height (same block execution mode, `ExecTxResult` as
+part of `PrepareProposal`).
+
+If `ExecTxResult` contains `Code != 0`, the transaction will be considered invalid,
+though it is still included in the block.
+
+`Events` include any events for the execution, which Tendermint will use to index
+the transaction by. This allows transactions to be queried according to what
+events took place during their execution.
+
+### Updating the Validator Set
+
+The application may set the validator set during
+[`InitChain`](./abci%2B%2B_methods_002_draft.md#initchain), and may update it during
+[`FinalizeBlock`](./abci%2B%2B_methods_002_draft.md#finalizeblock)
+(next block execution mode) or
+[`PrepareProposal`](./abci%2B%2B_methods_002_draft.md#prepareproposal)/[`ProcessProposal`](./abci%2B%2B_methods_002_draft.md#processproposal)
+(same block execution mode). In all cases, a structure of type
+[`ValidatorUpdate`](./abci%2B%2B_methods_002_draft.md#validatorupdate) is returned.
+
+The `InitChain` method can return a list of validators.
+If the list is empty, Tendermint will use the validators loaded in the genesis
+file.
+If the list returned by `InitChain` is not empty, Tendermint will use its contents as the validator set.
+This way the application can set the initial validator set for the
+blockchain.
+
+Note that the maximum total power of the validator set is bounded by
+`MaxTotalVotingPower = MaxInt64 / 8`. Applications are responsible for ensuring
+they do not make changes to the validator set that cause it to exceed this
+limit.
+
+Additionally, applications must ensure that a single set of updates does not contain any duplicates &mdash;
+a given public key can only appear once within a given update. If an update includes
+duplicates, the block execution will fail irrecoverably.
+
+Structure `ValidatorUpdate` contains a public key to reference the validator:
+
+The public key currently supports three types:
+
+* `ed25519`
+* `secp256k1`
+* `sr25519`
+
+Structure `ValidatorUpdate` also contains an `ìnt64` with the validator's new power.
+It must abide by the
+following rules:
+
+* power must be non-negative
+* if power is 0, the validator must already exist, and will be removed from the
+  validator set
+* if power is greater than 0:
+    * if the validator does not already exist, it will be added to the validator
+      set with the given power
+    * if the validator does already exist, its power will be adjusted to the given power
+* the total power of the new validator set must not exceed `MaxTotalVotingPower`
+
+Note the updates returned in block `H` will only take effect at block `H+2`.
```

Comment:
Make it clear: they are returned at the end of the processing phase of the block at height X. They will take effect (the validator set?) from block X+Y. Add a link to where this is detailed.